x-pack/plugins/security_solution/public/common/components/assignees/assignees_apply_panel.tsx

```diff
@@ -0,0 +1,156 @@
+    const { isLoading: isLoadingSuggestedUsers, data: userProfiles } = useSuggestUsers({
+      searchTerm,
+    });
+
+    const searchResultProfiles = useMemo(() => {
+      const sortedUsers = bringCurrentUserToFrontAndSort(currentUserProfile, userProfiles) ?? [];
+
+      if (showUnassignedOption && isEmpty(searchTerm)) {
+        return [NO_ASSIGNEES_VALUE, ...sortedUsers];
+      }
+
+      return sortedUsers;
+    }, [currentUserProfile, searchTerm, showUnassignedOption, userProfiles]);
+
+    const [selectedAssignees, setSelectedAssignees] = useState<AssigneesProfilesSelection[]>([]);
+    useEffect(() => {
+      if (isLoadingAssignedUsers || !assignedUsers) {
+        return;
+      }
+      const hasNoAssigneesSelection = assignedUserIds.find((uid) => uid === NO_ASSIGNEES_VALUE);
+      const newAssignees =
+        hasNoAssigneesSelection !== undefined
+          ? [NO_ASSIGNEES_VALUE, ...assignedUsers]
+          : assignedUsers;
+      setSelectedAssignees(newAssignees);
+    }, [assignedUserIds, assignedUsers, isLoadingAssignedUsers]);
+
+    const handleSelectedAssignees = useCallback(
+      (newAssignees: AssigneesProfilesSelection[]) => {
+        if (!isEqual(newAssignees, selectedAssignees)) {
+          setSelectedAssignees(newAssignees);
+          onSelectionChange?.(newAssignees.map((assignee) => assignee?.uid ?? NO_ASSIGNEES_VALUE));
+        }
+      },
+      [onSelectionChange, selectedAssignees]
+    );
+
+    const handleApplyButtonClick = useCallback(() => {
+      const selectedIds = selectedAssignees.map((assignee) => assignee?.uid ?? NO_ASSIGNEES_VALUE);
+      onAssigneesApply?.(selectedIds);
+    }, [onAssigneesApply, selectedAssignees]);
+
+    const selectedStatusMessage = useCallback(
+      (total: number) => i18n.ASSIGNEES_SELECTION_STATUS_MESSAGE(total),
+      []
+    );
+
+    const isLoading = isLoadingAssignedUsers || isLoadingSuggestedUsers;
+
+    return (
+      <div data-test-subj={ASSIGNEES_APPLY_PANEL_TEST_ID}>
+        <UserProfilesSelectable
+          searchInputId={searchInputId}
+          onSearchChange={(term: string) => {
+            setSearchTerm(term);
+          }}
+          onChange={handleSelectedAssignees}
+          selectedStatusMessage={selectedStatusMessage}
+          options={searchResultProfiles}
+          selectedOptions={selectedAssignees}
+          isLoading={isLoading}
+          height={'full'}
+          singleSelection={false}
+          searchPlaceholder={i18n.ASSIGNEES_SEARCH_USERS}
+          clearButtonLabel={i18n.ASSIGNEES_CLEAR_FILTERS}
+          nullOptionLabel={i18n.ASSIGNEES_NO_ASSIGNEES}
+        />
+        {onAssigneesApply && (
+          <EuiButton
+            data-test-subj={ASSIGNEES_APPLY_BUTTON_TEST_ID}
+            fullWidth
+            size="s"
+            onClick={handleApplyButtonClick}
+            isDisabled={isLoading}
+          >
+            {i18n.ASSIGNEES_APPLY_BUTTON}
+          </EuiButton>
+        )}
```

Comment:
I think this component would be clearer if the "Apply" button logic is extracted out. As I understand it, the `selectedAssignees` state is really only needed when `onAssigneesApply` is provided - otherwise, the `onChange` handler updates the selected profiles in some parent component and that parent component holds the real source of truth for selected assignee state. But, since `onAssigneesApply` is optional, it takes some extra effort to make that connection.

Perhaps this component could have the Apply button logic removed, and a new wrapper component could combine this component with the apply button. The wrapper would have require `onAssigneesApply`, and this component would then require `onSelectionChange`, instead of both being optional on this component. The wrapper component would maintain the selected assignee state (for the applying assignees to alerts use case), whereas for the "filtering by assignee" use case the `DetectionEnginePageComponent` would still maintain the selected assignee state as it does now. This way the source of truth for the selection state is always in the parent of this component. The prop types would also better communicate the intended usage of this component - right now it seems that either `on AssigneesApply` or `onSelectionChange` must be provided, but they're both optional. If we have a wrapper that requires `onAssigneesApply` and this component requires `onSelectionChange` then it's clearer how they're supposed to be used.